### PR TITLE
New js api

### PR DESCRIPTION
**Pull request by wacii**
Adds support for the drop-in, custom and advanced integrations of the Javascript SDK.

I don't know Sinatra, so there is likely a lot of poor code in the dummy app and serving of static assets. Also, I used the selenium driver for capybara instead of webkit as the latter wasn't working. Some problem with getting the braintree.js code to run/load, I think.

A few bits about the code not immediately obvious...

The client token encodes urls for the api, authorization, and assets servers that the javascript needs to run. It also encodes settings the user has selected in the dashboard/control panel/whatever, but I haven't addressed that here.

In the sinatra app, the route for `braintree.api.Client.prototype.getCreditCards()` is used in the drop-in integration to pre-populate the form. The braintree_assets folder is also used in the drop-in integration, as it normally loads all required files from the braintree servers.

### CODE REVIEW

**Comment by gabebw** [comment]
Whoa, thanks! This is a lot of work, thank you for doing it!

You can ignore the Hound comments - I can fix those, and I might have other feedback too.

Seriously, thanks!

**Comment by wacii** [comment]
Ah, no problem. Most of it is external libraries/assets needed by the the drop-in integration anyways.

Noticed the failing build. Travis CI not starting firefox. Seems the type of thing to be easily fixed, but I'll try and get the tests working with a headless driver first. Soonish.

**Comment by gabebw** [comment]
Adding this line to `.travis.yml` might run selenium:

```
script: xvfb-run bundle exec rake
```

File: lib/fake_braintree/braintree_assets/dropin/1.3.5/vendor/modernizr.js [hunk 1]
```diff
@@ -0,0 +1,4 @@
...
...tion(){try{var a=b.createElement("a");a.innerHTML="<xyz></xyz>",g="hidden"in a,k=a.childNodes.length==1||function(){b.createElement("a");var a=b.createDocumentFragment();return typeof a.cloneNode=="undefined"||typeof a.createDocumentFragment=="undefined"||typeof a.createElement=="undefined"}()}catch(c){g=!0,k=!0}})();var s={elements:d.elements||"abbr article aside audio bdi canvas data datalist details dialog figcaption figure footer header hgroup main mark meter nav output progress section summary template time video",version:c,shivCSS:d.shivCSS!==!1,supportsUnknownElements:k,shivMethods:d.shivMethods!==!1,type:"default",shivDocument:r,createElement:o,createDocumentFragment:p};a.html5=s,r(b)}(this,b),e._version=d,e._prefixes=n,e._domPrefixes=q,e._cssomPrefixes=p,e.testProp=function(a){return F([a])},e.testAllProps=H,e.testStyles=y,e.prefixed=function(a,b,c){return b?H(a,b,c):H(a,"pfx")},g.className=g.className.replace(/(^|\s)no-js(\s|$)/,"$1$2")+(f?" js "+v.join(" "):""),e}(this,this.document),Modernizr.addTest("placeholder",function(){return"placeholder"in(Modernizr.input||document.createElement("input"))&&"placeholder"in(Modernizr.textarea||document.createElement("textarea"))});
```
Comment (gabebw) [inline lib/fake_braintree/braintree_assets/dropin/1.3.5/vendor/modernizr.js]: Don't worry about it, Hound.

**Comment by twalpole** [comment]
Is there anything that can be done to get this merged sooner rather than later?

**Comment by birarda** [comment]
@wacii any chance you can take a look at the comments from @houndci?

I'd like to go through this and see what we need to do to get it merged.

**Comment by wacii** [comment]
Can you be more specific? Besides arguable that one long line, I thought I was following the repos conventions. Do you want me to change hound's settings?

**Comment by birarda** [comment]
It looks like there are some comments hound made that haven't been attended to yet - but maybe those are anomalies? I just started ownership of this repo today, I'll have to look through and see if those are inconsistent with the rest of the existing code.

**Comment by birarda** [comment]
Ahh, I see - @gabebw had said he would fix the hound stuff.

**Comment by wacii** [comment]
Hound was added recently. The settings never adapted to the project.

**Comment by gabebw** [comment]
Yes, wacii has a good point - this project is using single quotes everywhere, because that's how the project started and I never did a clean sweep of single -> double quotes.

**Comment by twalpole** [comment]
@birarda Is therre anything other than the hound warnings holding this up?  The single quote suggestions are pretty meaningless when one looks at the rest of the project, and this PR greatly increases the usefulness of this gem.

**Comment by birarda** [comment]
Agree @twalpole - the single quote stuff should not be a blocker. We've got a conference coming up next week so I am pretty blocked today. I haven't gotten a chance to even look through this at all.

I will give it a look on this upcoming Sunday or Monday so we can get this merged asap

**Comment by birarda** [comment]
I am going to start making my way through this today

**Pull request merged**